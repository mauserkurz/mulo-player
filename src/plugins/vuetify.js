import Vue from 'vue';
import Vuetify from 'vuetify';
import { THEME as theme } from '@/constants';

Vue.use(Vuetify);

export default new Vuetify({ theme });
