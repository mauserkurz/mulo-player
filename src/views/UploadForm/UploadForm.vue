<template>
  <v-card class="upload-form">
    <v-form
      ref="form"
      @submit.prevent="submit">
      <v-card-text>
        <v-file-input
          ref="file"
          :error="hasError"
          :rules="rules"
          accept="audio/mp3, audio/mpeg"
          placeholder="Select mp3 file"
          prepend-icon="mdi-file-music"
          label="Music file"
          @change="select"/>
      </v-card-text>

      <v-expand-transition>
        <v-card-text
          v-show="hasError">
          <v-alert
            type="error"
            dense>
            {{ error }}
          </v-alert>
        </v-card-text>
      </v-expand-transition>

      <v-card-actions>
        <v-spacer/>
        <v-btn
          :loading="loading"
          type="submit"
          color="primary">
          Save
        </v-btn>
      </v-card-actions>
    </v-form>
  </v-card>
</template>

<script>
import { MAX_FILE_SIZE } from '@/constants';

export default {
  name: 'UploadForm',

  props: {
    error: {
      type: String,
      default: '',
    },

    loading: {
      type: Boolean,
      default: false,
    },

    isHidden: {
      type: Boolean,
      default: false,
    },
  },

  data() {
    return {
      rules: [
        value => !!value || 'Please select mp3 file',
        value => !value || value.size < MAX_FILE_SIZE || 'File size should be less than 40 MB',
      ],
      file: null,
    };
  },

  computed: {
    hasError() {
      return !!this.error;
    },
  },

  watch: {
    isHidden(value) {
      if (value) {
        this.file = null;
        this.$refs.form.reset();
        this.$emit('clear-form');
      }
    },
  },

  methods: {
    select(file) {
      this.file = file;
    },

    submit() {
      if (this.$refs.form.validate()) {
        const typedFile = new File([this.file], this.file.name, { type: 'audio/mpeg' });

        this.$emit('submit', typedFile);
      }
    },
  },
};
</script>

<style scoped lang="less">
.upload-form {}
</style>
