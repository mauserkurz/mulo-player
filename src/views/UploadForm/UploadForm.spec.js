import { mount, shallowMount, createLocalVue } from '@vue/test-utils';
import { toMatchDiffSnapshot } from 'snapshot-diff';
import flushPromises from 'flush-promises';
import vuetify from '@/plugins/vuetify';
import UploadForm from './UploadForm.vue';

const localVue = createLocalVue();

expect.extend({ toMatchDiffSnapshot });

describe('Component UploadForm', () => {
  const mockFileInput = (input) => {
    let files;

    Object.defineProperty(input.element, 'files', {
      get() {
        return files;
      },
    });

    Object.defineProperty(input.element, 'value', {
      get() {
        return files;
      },

      set(value) {
        files = value;
      },
    });
  };

  const createWrapper = ({ isWithoutStubs = false, ...options } = {}) => {
    const renderer = isWithoutStubs ? mount : shallowMount;
    const wrapper = renderer(UploadForm, {
      localVue,
      vuetify,
      ...options,
    });

    return { wrapper };
  };

  describe('snapshots', () => {
    it('should match snapshot for login form', () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true });

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('props', () => {
    it('should show error message from prop', () => {
      const { wrapper } = createWrapper({ propsData: { error: 'Some error message' } });

      expect(wrapper.find('v-alert-stub').text()).toBe('Some error message');
    });

    it('should show spinner in submit button while loading', () => {
      const { wrapper } = createWrapper({ propsData: { loading: true } });

      expect(wrapper.find('v-btn-stub').vm.loading).toBe(true);
    });

    describe('isHidden represent drawer showing change', () => {
      it('should after change isHidden to true reset form', async () => {
        const { wrapper } = createWrapper({ isWithoutStubs: true, propsData: { isHidden: false } });
        const spy = jest.spyOn(wrapper.vm.$refs.form, 'reset');

        wrapper.setProps({ isHidden: true });
        await wrapper.vm.$nextTick();
        expect(spy).toHaveBeenCalled();
      });

      it('should after change isHidden to true send event clear-form', async () => {
        const { wrapper } = createWrapper({ isWithoutStubs: true, propsData: { isHidden: false } });

        wrapper.setProps({ isHidden: true });
        await wrapper.vm.$nextTick();
        expect(wrapper.emitted()['clear-form']).toEqual([[]]);
      });
    });
  });

  describe('events', () => {
    it('should show required error for file field', async () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true });

      wrapper.find('.v-form').vm.$emit('submit', { preventDefault() {} });
      await flushPromises();
      expect(wrapper.findAll('.v-messages__message').at(0).text())
        .toBe('Please select mp3 file');
    });

    it('should not emit submit event for file field with error', async () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true });

      wrapper.find('.v-form').vm.$emit('submit', { preventDefault() {} });
      await flushPromises();
      expect(wrapper.emitted().submit).toBeUndefined();
    });

    it('should emit submit for valid form with changed meme type', async () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true });
      const input = wrapper.find('input[type="file"]');
      const file = new File([], 'track', { type: 'audio/mp3' });

      mockFileInput(input);
      input.element.value = [file];
      input.trigger('change');

      wrapper.find('.v-form').vm.$emit('submit', { preventDefault() {} });
      await flushPromises();
      expect(wrapper.emitted().submit).toEqual([[file]]);
    });
  });
});
