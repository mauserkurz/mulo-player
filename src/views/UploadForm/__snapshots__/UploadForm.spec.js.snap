// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component UploadForm snapshots should match snapshot for login form 1`] = `
<div class="upload-form v-card v-sheet theme--light">
  <form novalidate="novalidate" class="v-form">
    <div class="v-card__text">
      <div class="v-input theme--light v-text-field v-text-field--placeholder v-file-input">
        <div class="v-input__prepend-outer">
          <div class="v-input__icon v-input__icon--prepend"><i role="button" tabindex="0" class="v-icon notranslate v-icon--link mdi mdi-file-music theme--light"></i></div>
        </div>
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot"><label for="input-5" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Music file</label>
              <div class="v-file-input__text v-file-input__text--placeholder">Select mp3 file</div><input accept="audio/mp3, audio/mpeg" id="input-5" placeholder="Select mp3 file" type="file">
            </div>
            <div class="v-input__append-inner">
              <div class="v-input__icon v-input__icon--"><i role="button" tabindex="0" class="v-icon notranslate v-icon--link material-icons theme--light"></i></div>
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
          </div>
        </div>
      </div>
    </div>
    <div class="v-card__text" style="display: none;">
      <div role="alert" class="v-alert v-sheet theme--dark v-alert--dense error">
        <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-alert theme--dark"></i>
          <div class="v-alert__content">

          </div>
        </div>
      </div>
    </div>
    <div class="v-card__actions">
      <div class="spacer"></div> <button type="submit" class="v-btn v-btn--contained theme--light v-size--default primary"><span class="v-btn__content">
        Save
      </span></button>
    </div>
  </form>
</div>
`;
